### PR TITLE
v0.6

### PR DESCRIPTION
**Pull request by StRigaud**
Rework of core architecture and backend usage

- encapsulation of opencl code into backend function :arrow_right:  `cle::Backend`
- memory allocation, write, and read dissociated from main gpu class and implemented as function :arrow_right:  `cle::Memory`
- renaming of major classes: gpu -> processor, object -> image
- usage of minimal struct object for data type and object type management :arrow_right:  `cleUtils.hpp`

Smaller improvement

- Image object can do data copy :arrow_right: `image.CoptyDataTo( other image)`
- Image object can buffer filled :arrow_right: `image.Fill(value)`

Code clarity

- enforce usage of clang-format and clang-tidy for improved code quality

### CODE REVIEW

**Comment by StRigaud** [comment]
Reaching quite a stable build here.

Most of the main correction and improvement points are done. There still some improvement in code clarity and simplification but this should not change the architecture and can be done later on with minor patches.

I will let it sit for a bit and then merge it. Updates on the python and java wrapper can start.

File: clic/include/core/cleImage.hpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  [[nodiscard]] auto
+  DataInfoShort() const -> std::string override;
+  [[nodiscard]] auto
+  IsBuffer() const -> bool;
+  [[nodiscard]] auto
+  IsImage() const -> bool;
+  [[nodiscard]] auto
+  ToString() const -> std::string override;
+  [[nodiscard]] auto
+  Data() const -> DataType override;
```
Comment (haesleinhuepf) [inline clic/include/core/cleImage.hpp:62]: Might it makes sense to call this `DataType` and not `Data`? I just hit this while updating code...

Comment (StRigaud) [reply clic/include/core/cleImage.hpp:62]: agreed, renaming methods to `DataType` and `ObjectType` is planned
return type will be rename `DataTypeEnum` etc.

Comment (StRigaud) [reply clic/include/core/cleImage.hpp]: Also, I moved methods to the abstract class `LightObject` to reduce code replication and increase class clarity

File: clic/src/core/clesperanto.cpp [hunk 2]
```diff
@@ -458,9 +458,7 @@ auto
 Clesperanto::Absolute(const Image & source, const Image & destination) -> void
 {
   AbsoluteKernel kernel(this->GetDevice());
-  kernel.SetInput(source);
-  kernel.SetOutput(destination);
-  kernel.Execute();
+  Absolute_call(&kernel, source, destination);
 }
```
[the remark this answers is not in the document]
Comment (StRigaud) [reply clic/src/core/clesperanto.cpp]: I've update the idea with a simple "inline function" instead of a "friend function".
The idea remain the same.

**Comment by StRigaud** [comment]
Ready to Merge

Final look about retrograding to cxx14 std for compatibility and we should be good :love_you_gesture:

**Pull request merged**